### PR TITLE
fix: support for deeply nested `format: json` schemas

### PR DESCRIPTION
**Pull request by erunion**
| 🚥 Fixes CX-229 |
| :---------------- |

## 🧰 Changes

This fixes a quirk in our handling of schemas with `format: json` where we only supported them if they were at the top-most level in the schema. If you had an object with one of its properties being `format: json`, or an array of objects with the same, that JSON object would be stringified into the HAR we generate.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -544,16 +548,12 @@ export default function oasToHar(
...
-            /**
-             * Handle arbitrary JSON input via a string.
-             *
-             * In OAS you usually find this in an `application/json` content type with a schema
-             * `type=string, format=json`. In the UI this is represented by an arbitrary text input.
-             *
-             * This ensures we remove any newlines or tabs and use a clean JSON block in the
-             * example.
-             */
-            if ((requestBody.schema as SchemaObject).type === 'string') {
```
Comment (erunion) [inline src/index.ts:556]: This didn't account for if OpenAPI 3.1's support for mixed types.

File: src/index.ts [hunk 2]
```diff
@@ -430,10 +433,11 @@ export default function oasToHar(
...
   }
 
-  let requestBody: MediaTypeObject;
+  let requestBody: SchemaWrapper;
   if (operation.hasRequestBody()) {
-    // @ts-expect-error TODO `requestBody` coming back as `false | MediaTypeObject | [string, MediaTypeObject]` seems like a problem
-    [, requestBody] = operation.getRequestBody();
+    requestBody = operation
+      .getParametersAsJSONSchema()
+      .find(p => p.type === (operation.isFormUrlEncoded() ? 'formData' : 'body'));
```
Comment (erunion) [inline src/index.ts:440]: `formData` is used in our API Explorer for `application/x-www-form-urlencoded` endpoints and if you have an operation with that, it will only ever have a `formData`. `body` is used for all other payload shapes.

File: src/lib/utils.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  return schema.type === discriminator;
+}
+
+/**
+ * With a supplied JSON Schema object, spider through it for any schemas that may contain specific
+ * kind of `format` that also happen to be within the current `requestBody` payload that we're
+ * creating a HAR representation for.
+ *
+ */
+export function getTypedFormatsInSchema(
```
Comment (erunion) [inline src/lib/utils.ts:26]: With the test I wrote in `requestBody.test.ts` this thing ends up returning an array that looks like:

```js
['root', 'obj.prop', 'arr.0', 'arrOfObjects.0.arrProp', 'arrOfObjects.1.arrProp']
```

And because that's dot-notation we can easily use Lodash `get` and `set` to manipulate that data into JSON objects.

File: src/lib/utils.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  }
+): boolean | string | (string | boolean)[] {
+  try {
+    if (schema?.format === format) {
+      if (opts.parentIsArray) {
+        const parentData = lodashGet(opts.payload, opts.parentKey);
+        if (parentData !== undefined && Array.isArray(parentData)) {
+          return Object.keys(parentData)
+            .map(pdk => {
+              const currentKey = [opts.parentKey, pdk /* , key */].join('.');
```
Comment (ilias-t) [inline src/lib/utils.ts:42]: I don't really understand the `key` comment, is that just an artifact of development?

File: src/lib/utils.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+ * creating a HAR representation for.
+ *
+ */
+export function getTypedFormatsInSchema(
+  format: 'json' | 'binary',
+  schema: any,
+  opts: {
+    parentIsArray?: boolean;
+    parentKey?: string;
+    payload: any;
```
Comment (ilias-t) [inline src/lib/utils.ts:32]: nit
```suggestion
    payload: unknown;
```

File: src/lib/utils.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+ * With a supplied JSON Schema object, spider through it for any schemas that may contain specific
+ * kind of `format` that also happen to be within the current `requestBody` payload that we're
+ * creating a HAR representation for.
+ *
+ */
+export function getTypedFormatsInSchema(
+  format: 'json' | 'binary',
+  schema: any,
+  opts: {
+    parentIsArray?: boolean;
```
Comment (ilias-t) [inline src/lib/utils.ts:30]: Could this not be inferred via the `parentKey`?

**Review by ilias-t — APPROVED** [review]

File: src/lib/utils.ts (hunk 4 above)
Comment (erunion) [reply src/lib/utils.ts:42]: Artifact yeah, will remove.

File: src/lib/utils.ts (hunk 6 above)
Comment (erunion) [reply src/lib/utils.ts:30]: Unfortunately not because `parentKey` could be a parent key for an object, or a regular schema.

File: src/lib/utils.ts (hunk 3 above)
Comment (mosesj-readme) [inline src/lib/utils.ts:26]: this block of code is hard to parse through. maybe breaking it up into functions will make it easier to read.

```suggestion

function handleParentDataArrays(parentData: any, opts: { parentIsArray?: boolean; parentKey?: string; payload: any }) {
  return Object.keys(parentData)
    .map(pdk => {
      const currentKey = [opts.parentKey, pdk /* , key */].join('.');
      if (lodashGet(opts.payload, currentKey) !== undefined) {
        return currentKey;
      }

      return false;
    })
    .filter(Boolean);
}

function getSubSchemas(subSchemaDataSize: number, schema: any, opts: {
  parentIsArray?: boolean;
  parentKey?: string;
  payload: any
}) {
  let subSchemas: any[] = [];
  if (subSchemaDataSize > 0) {
    for (let idx = 0; idx < subSchemaDataSize; idx += 1) {
      subSchemas = subSchemas.concat(
        Object.entries(schema).map(([key, subSchema]: [string, JSONSchema]) => ({
          key: opts.parentKey ? [opts.parentKey, idx, key].join('.') : key,
          schema: subSchema,
        }))
      );
    }
  } else {
    return Object.entries(schema).map(([key, subSchema]: [string, JSONSchema]) => ({
      key: opts.parentKey ? [opts.parentKey, key].join('.') : key,
      schema: subSchema,
    }));
  }
}

/**
 * With a supplied JSON Schema object, spider through it for any schemas that may contain specific
 * kind of `format` that also happen to be within the current `requestBody` payload that we're
 * creating a HAR representation for.
 *
 */
export function getTypedFormatsInSchema(
  format: 'json' | 'binary',
  schema: any,
  opts: {
    parentIsArray?: boolean;
    parentKey?: string;
    payload: any;
  }
): boolean | string | (string | boolean)[] {
  try {
    if (schema?.format === format) {
      const parentData = lodashGet(opts.payload, opts.parentKey);
      if (opts.parentIsArray && parentData !== undefined && Array.isArray(parentData)) {
        return handleParentDataArrays(parentData, opts);
      } else if (opts.parentKey && parentData !== undefined) {
        return opts.parentKey;
      } else if (opts.payload !== undefined) {
        // If this payload is present and we're looking for a specific format then we should assume
        // that the **root** schema of the request body is that format, and we aren't trafficking in
        // a nested object or array schema.
        return true;
      }

      return false;
    }

    let subSchemaDataSize = 0;
    if (opts.parentIsArray) {
      // If we don't have data for this parent schema in our body payload then we
      // shouldn't bother spidering further into the schema looking for more `format`s
      // for data that definitely doesn't exist.
      const parentData = lodashGet(opts.payload, opts.parentKey);
      if (parentData === undefined || !Array.isArray(parentData)) {
        return false;
      }

      subSchemaDataSize = parentData.length;
    }
    const subSchemas = getSubSchemas(subSchemaDataSize, schema, opts);

    return subSchemas
      .map(({key, schema: subSchema}) => {
        if ('properties' in subSchema) {
          return getTypedFormatsInSchema(format, subSchema.properties, {payload: opts.payload, parentKey: key});
        }

        if ('items' in subSchema) {
          const schema = (subSchema.items as JSONSchema)?.properties ?? subSchema.items;

          return getTypedFormatsInSchema(format, schema, {
            payload: opts.payload,
            parentKey: key,
            parentIsArray: true,
          });
        }

        // If this schema has neither `properties` or `items` then it's a regular schema we can re-run.
        return getTypedFormatsInSchema(format, subSchema, {payload: opts.payload, parentKey: key});
      })
      .flat()
      .filter(Boolean);
  } catch (err) {
    // If this fails for whatever reason then we should act as if we didn't find any `format`'d
    // schemas.
    return [];
  }
}
```

**Review by mosesj-readme — APPROVED** [review]
left comments but not going to block for them

File: src/index.ts (hunk 2 above)
Comment (erunion) [reply src/index.ts:440]: sure

File: src/lib/utils.ts (hunk 3 above)
Comment (erunion) [reply src/lib/utils.ts:26]: Good idea
